orangecontrib/prototypes/widgets/contingency_table.py

```diff
@@ -0,0 +1,365 @@
+        self.setItemDelegateForColumn(0, gui.VerticalItemDelegate())
+        self._set_item(2, 0, item)
+        self.setSpan(0, 2, 1, len(self.classesh)+1)
+        self.setSpan(2, 0, len(self.classesv)+1, 1)
+
+        font = self.tablemodel.invisibleRootItem().font()
+        bold_font = QFont(font)
+        bold_font.setBold(True)
+
+        for i in (0, 1):
+            for j in (0, 1):
+                item = self._item(i, j)
+                item.setFlags(Qt.NoItemFlags)
+                self._set_item(i, j, item)
+
+        for headers, ix in ((self.classesv + [self.corner_string], lambda p: (p + 2, 1)),
+                            (self.classesh + [self.corner_string], lambda p: (1, p + 2))):
+            for p, label in enumerate(headers):
+                i, j = ix(p)
+                item = self._item(i, j)
+                item.setData(label, Qt.DisplayRole)
+                item.setFont(bold_font)
+                if not (i == 1 and self.circles):
+                    item.setTextAlignment(Qt.AlignRight | Qt.AlignVCenter)
+                item.setFlags(Qt.ItemIsEnabled)
+                if p < len(headers) - 1:
+                    item.setData("br"[j == 1], BorderRole)
+                    item.setData(QColor(192, 192, 192), BorderColorRole)
+                else:
+                    item.setData("", BorderRole)
+                self._set_item(i, j, item)
+
+        hor_header = self.horizontalHeader()
+        if len(' '.join(self.classesh + [self.corner_string])) < 120:
+            hor_header.setSectionResizeMode(QHeaderView.ResizeToContents)
+        else:
+            hor_header.setDefaultSectionSize(60)
+        if self.circles:
+            self.resizeRowToContents(1)
+            self.tablemodel.setRowCount(len(self.classesv) + 2)
+            self.tablemodel.setColumnCount(len(self.classesh) + 2)
+        else:
+            self.tablemodel.setRowCount(len(self.classesv) + 3)
+            self.tablemodel.setColumnCount(len(self.classesh) + 3)
+
+    def get_selection(self):
+        """
+        Get indexes of selected cells.
+
+        Returns
+        -------
+        :obj:`set` of :obj:`tuple` of :obj:`int`
+            Set of pairs of indexes.
+        """
+        return {(ind.row() - 2, ind.column() - 2) for ind in self.selectedIndexes()}
+
+    def set_selection(self, indexes):
+        """
+        Set indexes of selected cells.
+
+        Parameters
+        ----------
+        indexes : :obj:`set` of :obj:`tuple` of :obj:`int`
+            Set of pairs of indexes.
+        """
+        selection = QItemSelection()
+        index = self.model().index
+        for row, col in indexes:
+            sel = index(row + 2, col + 2)
+            selection.select(sel, sel)
+        self.selectionModel().select(
+            selection, QItemSelectionModel.ClearAndSelect)
+
+    def update_table(self, matrix, colsum=None, rowsum=None, colors=None, formatstr="{}", tooltip=None):
+        """
```

Comment:
This method could also be split into several shorter (private) methods. (i'd go with at least one method for creating cells, one for sums).